core/darwin/src/internal/TimeZoneRulesFoundation.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2019-2025 JetBrains s.r.o. and contributors.
+ * Use of this source code is governed by the Apache 2.0 License that can be found in the LICENSE.txt file.
+ */
+
+package kotlinx.datetime.internal
+
+import kotlinx.cinterop.UnsafeNumber
+import kotlinx.datetime.LocalDateTime
+import kotlinx.datetime.UtcOffset
+import kotlinx.datetime.toKotlinInstant
+import kotlinx.datetime.toNSDate
+import kotlinx.datetime.toNSDateComponents
+import platform.Foundation.NSCalendar
+import platform.Foundation.NSDate
+import platform.Foundation.NSTimeZone
+import platform.Foundation.addTimeInterval
+import platform.Foundation.timeZoneWithName
+import kotlin.time.Instant
+
+internal class TimeZoneRulesFoundation(zoneId: String) : TimeZoneRules {
+    private val nsTimeZone: NSTimeZone = NSTimeZone.timeZoneWithName(zoneId)
+        ?: throw IllegalArgumentException("Unknown timezone: $zoneId")
+
+    override fun infoAtInstant(instant: Instant): UtcOffset =
+        infoAtNsDate(instant.toNSDate())
+
+    override fun infoAtDatetime(localDateTime: LocalDateTime): OffsetInfo {
+        val calendar = NSCalendar.currentCalendar().apply { timeZone = nsTimeZone }
```

Comment:
Using `currentCalendar` will mean that the behavior will be different depending on the user's system configuration. This is an approach we are explicitly avoiding in `kotlinx-datetime`: for example, in java.time, the system timezone or locale is occasionally queried implicitly, and we don't do that.
Instead, this calendar can be used: https://developer.apple.com/documentation/foundation/nscalendar/identifier/iso8601